### PR TITLE
Stop assuming encrypted messages are decrypted ahead of time

### PR DESCRIPTION
**Pull request by gsouquet**
To review alongside matrix-org/matrix-js-sdk#1696

### CODE REVIEW

File: src/components/structures/FilePanel.js [hunk 1]
```diff
@@ -50,6 +50,10 @@ class FilePanel extends React.Component {
         if (room?.roomId !== this.props?.roomId) return;
         if (toStartOfTimeline || !data || !data.liveEvent || ev.isRedacted()) return;
 
+        if (ev.shouldAttemptDecryption()) {
+            ev.attemptDecryption(room._client._crypto);
```
Comment (t3chguy) [inline src/components/structures/FilePanel.js:54]: we shouldn't be accessing private members like this, there's no guarantee they won't change between releases

File: src/components/views/messages/MessageActionBar.js [hunk 2]
```diff
@@ -122,6 +123,12 @@ export default class MessageActionBar extends React.PureComponent {
         if (this.props.mxEvent.status && this.props.mxEvent.status !== EventStatus.SENT) {
             this.props.mxEvent.on("Event.status", this.onSent);
         }
+
+        if (this.props.mxEvent.shouldAttemptDecryption()) {
+            const client = MatrixClientPeg.get();
+            this.props.mxEvent.attemptDecryption(client._crypto);
```
Comment (t3chguy) [inline src/components/views/messages/MessageActionBar.js:129]: same as above

File: src/components/views/messages/ViewSourceEvent.js [hunk 3]
```diff
@@ -36,6 +37,12 @@ export default class ViewSourceEvent extends React.PureComponent {
 
     componentDidMount() {
         const {mxEvent} = this.props;
+
+        const client = MatrixClientPeg.get();
+        if (mxEvent.shouldAttemptDecryption()) {
+            mxEvent.attemptDecryption(client._client._crypto);
```
Comment (t3chguy) [inline src/components/views/messages/ViewSourceEvent.js:43]: same as above

File: src/indexing/EventIndex.js [hunk 4]
```diff
@@ -187,6 +187,10 @@ export default class EventIndex extends EventEmitter {
             return;
         }
 
+        if (ev.shouldAttemptDecryption()) {
+            ev.attemptDecryption(room._client._crypto);
```
Comment (t3chguy) [inline src/indexing/EventIndex.js:191]: same as above

**Review by t3chguy — CHANGES_REQUESTED** [review]

**Comment by jryans** [comment]
Perhaps one way forward is for the JS SDK `client` object to gain its own `attemptDecryption` that takes an event, and then it supplies its private crypto field to the event's `attemptDecryption`?

The JS SDK design overall needs some rethinking (we just keeping forwarding methods from crypto onto the client) some day, but that's for another time. Anyway, something like that is probably reasonable given current practice.

**Comment by jryans** [comment]
@gsouquet With that new client API added, we should also remember to clean up the existing calls already added in `EventIndex` and `TimelinePanel`.

File: src/components/structures/FilePanel.js [hunk 5]
```diff
@@ -50,6 +50,9 @@ class FilePanel extends React.Component {
         if (room?.roomId !== this.props?.roomId) return;
         if (toStartOfTimeline || !data || !data.liveEvent || ev.isRedacted()) return;
 
+        const client = MatrixClientPeg.get();
+        client.decryptEvent(ev);
```
Comment (t3chguy) [inline src/components/structures/FilePanel.js:54]: seems weird to call a method called `decryptEvent` on something which may be not encrypted in the first place, either needs wrapping in an if check of `shouldAttemptDecryption` or the method could be renamed something like `decryptEventIfNeeded`

File: src/components/views/messages/MessageActionBar.js [hunk 6]
```diff
@@ -122,6 +123,12 @@ export default class MessageActionBar extends React.PureComponent {
         if (this.props.mxEvent.status && this.props.mxEvent.status !== EventStatus.SENT) {
             this.props.mxEvent.on("Event.status", this.onSent);
         }
+
+        if (this.props.mxEvent.shouldAttemptDecryption()) {
```
Comment (t3chguy) [inline src/components/views/messages/MessageActionBar.js:127]: this if check is done inside `decryptEvent` too

File: src/components/structures/TimelinePanel.js [hunk 7]
```diff
@@ -1150,7 +1150,8 @@ class TimelinePanel extends React.Component {
             .reverse()
             .forEach(event => {
                 if (event.shouldAttemptDecryption()) {
```
Comment (t3chguy) [inline src/components/structures/TimelinePanel.js:1152]: this is already done inside `decryptEvent`

File: src/components/views/messages/ViewSourceEvent.js [hunk 8]
```diff
@@ -36,6 +37,12 @@ export default class ViewSourceEvent extends React.PureComponent {
 
     componentDidMount() {
         const {mxEvent} = this.props;
+
+        const client = MatrixClientPeg.get();
+        if (mxEvent.shouldAttemptDecryption()) {
```
Comment (t3chguy) [inline src/components/views/messages/ViewSourceEvent.js:42]: `shouldAttemptDecryption` gets done inside decryptEvent too

File: src/indexing/EventIndex.js [hunk 9]
```diff
@@ -187,10 +189,7 @@ export default class EventIndex extends EventEmitter {
             return;
         }
 
-        if (ev.isBeingDecrypted()) {
-            // XXX: Private member access
-            await ev._decryptionPromise;
-        }
+        await client.decryptEvent(ev);
```
Comment (t3chguy) [inline src/indexing/EventIndex.js:192]: method name seems misleading given the event may be plaintext

**Review by t3chguy — APPROVED** [review]
I think this is generally fine but the signature/naming might need tweaking

**Review by jryans — CHANGES_REQUESTED** [review]
Overall, this looks good, but there may be a few more call sites to check:

https://github.com/matrix-org/matrix-react-sdk/blob/367ad1583d79eccf79330d14edf9ac3ece196646/src/Notifier.ts#L336-L343

https://github.com/matrix-org/matrix-react-sdk/blob/367ad1583d79eccf79330d14edf9ac3ece196646/src/stores/widgets/StopGapWidget.ts#L403

**Review by jryans — APPROVED** [review]
Thanks, I think this is ready to go now. 😄

**Pull request merged**